File: Aeds/AedsII/Atividades/Trabalhos Praticos/Tp3/Num1/codigo.java
import java.io.BufferedReader;
import java.io.FileReader;


class Jogador {

    /*
     * Aluno(Dev): Tulio Gomes Braga;
     * Usuario: 1441272;
     * 
     * Descrição do conteudo:
     * 
     * Atributos da classe jogador, seguindo a tabela "players.csv";
     * Metodos sets e gets para manipular os valores do objeto;
     * Construtor: esse metodo fica responsavel por cria um objeto com valores nulos;
     * Clone: objeto que chama tem seus atributos copiados pra uma nova instância de objeto;
     * Inserir Elementos: Recebe valores ate ser igual a "FIM", cria um novo objeto, tenta transformar em inteiro e chama os proximos metodos;
     * SetaId: Recebe um id, abre a leitura no arquivo e vai percorrendo até o id lido seja igual ao recebido, caso for pega a linha correspondente;
     * SetaDados: Recebe uma linha, fragmenta de acordo com a virgula e insere os argumentos daquele objeto corrente;
     * MostraDados: Mostra todos os atributos do objeto corrente;
     * MostrarLista: Metodo pra percorrer toda nossa lista e printar os dados;
     * Escreve: Recebe o objeto corrente e printa o retorno de outro metodo;
     * toString: Similar ao mostrar dados, retorna os dados concatenados em String;
     * Main: 
     * 
     * Exercicio:
    */

    public static int tamanho=0, MAX=130;
    private Integer id,altura,peso,anoNascimento;
    private String nome,universidade,cidadeNascimento,estadoNascimento;

    public static Integer comp = 0;
    public static Long tempo = (long) 0;

    public String getEstadoNascimento() {
        return estadoNascimento;
    }
    public String getCidadeNascimento() {
        return cidadeNascimento;
    }
    public String getUniversidade() {
        return universidade;
    }
    public String getNome() {
        return nome;
    }
    public int getAnoNascimento() {
        return anoNascimento;
    }
    public int getPeso() {
        return peso;
    }
    public int getAltura() {
        return altura;
    }
    public int getId() {
        return id;
    }
    public void setId(int id) {
        this.id = id;
    }
    public void setAltura(int altura) {
        this.altura = altura;
    }
    public void setPeso(int peso) {
        this.peso = peso;
    }
    public void setAnoNascimento(int anoNascimento) {
        this.anoNascimento = anoNascimento;
    }
    public void setNome(String nome) {
        this.nome = nome;
    }
    public void setUniversidade(String universidade) {
        this.universidade = universidade;
    }
    public void setCidadeNascimento(String cidadeNascimento) {
        this.cidadeNascimento = cidadeNascimento;
    }
    public void setEstadoNascimento(String estadoNascimento) {
        this.estadoNascimento = estadoNascimento;
    }

    public Jogador(){
        this.id = null;
		this.nome = null;
		this.altura = null;
		this.peso = null;
		this.universidade = null;
		this.anoNascimento = null;
		this.cidadeNascimento = null;
		this.estadoNascimento = null;
    }

    public Jogador cloneJogador() {
		Jogador clone = new Jogador();
		clone.id = this.getId();
		clone.nome = this.getNome();
		clone.altura = this.getAltura();
		clone.peso = this.getPeso();
		clone.universidade = this.getUniversidade();
		clone.anoNascimento = this.getAnoNascimento();
		clone.cidadeNascimento = this.getCidadeNascimento();
		clone.estadoNascimento = this.getEstadoNascimento();
		return clone;
	}

    /*---------------------------------------------------------------------------------------------------------------------------------------------------------------- INSERCAO DE DADOS INICIO */

    public static void inserirElementos(Jogador[] lista){
        String input = MyIO.readLine();
        while (!input.equals("FIM")) {
            try {
                Integer id = Integer.parseInt(input);
                Jogador jogador = new Jogador();
                jogador.SetaId(id);
                lista[tamanho] = jogador;
                tamanho++;
            } catch (Exception e) {}
            input = MyIO.readLine();
        }
        
    }

    public void SetaId(int x) throws Exception {
        FileReader file = new FileReader("players.csv"); 
        BufferedReader buffer = new BufferedReader(file);
        String linha;

        while ((linha = buffer.readLine()) != null) {
            String numId = linha.substring(0, linha.indexOf(","));

            try{
                if(Integer.parseInt(numId) == x) {
                    this.setaDados(linha);
                    //this.mostraDados();
                    break;
                }
            }catch(NumberFormatException e){}
        }
        buffer.close();
        file.close();
    }

    public void setaDados(String linha){
        String[] parte = linha.split(",");

        this.setId(Integer.parseInt(parte[0]));
        this.setNome(parte[1]);
        this.setAltura(Integer.parseInt(parte[2]));
        this.setPeso(Integer.parseInt(parte[3]));
        if(parte.length>4 && !parte[4].isEmpty()) this.setUniversidade(parte[4]); else this.setUniversidade("nao informado");
        this.setAnoNascimento(Integer.parseInt(parte[5]));
        if (parte.length>6 && !parte[6].isEmpty()) this.setCidadeNascimento(parte[6]); else this.setCidadeNascimento("nao informado");
        if (parte.length>7 && !parte[7].isEmpty()) this.setEstadoNascimento(parte[7]); else this.setEstadoNascimento("nao informado");

    }

    public void mostraDados(){
        MyIO.println("[" + this.getId() + " ## " + this.getNome() + " ## " + this.getAltura() + " ## " + this.getPeso() + " ## " + this.getAnoNascimento() + " ## " 
        + this.getUniversidade() + " ## " + this.getCidadeNascimento() + " ## " + this.getEstadoNascimento() + "]");
    }

    public static void MostraLista(Jogador[] lista) {
		for(int i = 0; i < tamanho; i++) {
			lista[i].Escrever();
		}
	}

    public void Escrever() {
		String linha = this.toString();
		MyIO.println(linha);
	}

    public String toString() {
		String txt = "";
		txt += this.getId() + " ## ";
		txt += this.getNome() + " ## ";
		txt += this.getAltura() + " ## ";
		txt += this.getPeso() + " ## ";
		txt += this.getAnoNascimento() + " ## ";
		txt += this.getUniversidade() + " ## ";
		txt += this.getCidadeNascimento() +  " ## ";
		txt += this.getEstadoNascimento() + "#";
		return txt;
	}

    /*---------------------------------------------------------------------------------------------------------------------------------------------------------------- INSERCAO DE DADOS FIM */
    /*---------------------------------------------------------------------------------------------------------------------------------------------------------- MANIPULACAO DE DADOS INICIO */

    public static void manipulaElementos(Jogador[] lista){
        int qtd = MyIO.readInt();

        for(int i=0;i<qtd;i++){
            String input = MyIO.readLine();
            decodificaInput(input,lista);
        }
    }

    public static void decodificaInput(String input, Jogador[] lista){
        /*
         * II inserir no início
         * I* inserir em qualquer posição
         * IF inserir no fim
         * RI remover no início
         * R* remover em qualquer posição
         * RF remover no fim.
         */
        String[] parte = input.split(" ");

        if(parte.length == 1){
            switch (parte[0]){ 
                case "RI": removerInicio(); break; 
                case "RF": removerFim(); break; 
            }
        }
        else if(parte.length == 2){
            int idJogador =  Integer.parseInt(parte[1]);
            switch (parte[0]){
                case "II": inserirInicio(idJogador); break;     
                case "IF": inserirFim(idJogador); break; 
                case "R*": remover(idJogador); break;  
            }
        }
        else if(parte.length == 3){
            int posicao =  Integer.parseInt(parte[1]);
            int idJogador =  Integer.parseInt(parte[2]);
            inserir(idJogador, posicao);
        }

    }

    public static void inserirInicio(int idJogador){
        MyIO.println("Inserir inicio");
    }

    public static void inserir(int idJogador, int posicao){
         MyIO.println("Inserir");
    }

    public static void inserirFim(int idJogador){
         MyIO.println("Inserir Fim");
    }

    public static void removerInicio(){
         MyIO.println("Remover inicio");
    }

    public static void remover(int posicao){
         MyIO.println("Remover");
    }

    public static void removerFim(){
         MyIO.println("Remover Fim");
    }


    public static void main(String[] args) {            // Compilar:    javac .\Jogador.java 
        Jogador[] array = new Jogador[MAX];             // Executar:    type .\pub.in | java Jogador > result.txt */

        inserirElementos(array);
        manipulaElementos(array);
        //MostraLista(array);

    }

}

File: Aeds/AedsII/Atividades/Trabalhos Praticos/Tp3/Num2/codigo.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

/*
    Dev: Tulio Gomes Braga
    Matricula: 1441272
    Curso: CC-PucMinas - 13/10/2023
*/

int tamanho = 0, mostrar=0;

typedef struct {
    int id;
    int altura;
    int peso;
    int anoNascimento;
    char nome[100];
    char universidade[100];
    char cidadeNascimento[100];
    char estadoNascimento[100];
} Jogador;


void inserirElementos(Jogador *lista);
void setaId(Jogador *jogador, int id);
void parteDados(Jogador *jogador, const char *linha);
void mostraDados(Jogador *jogador);
void mostrarTodos(Jogador *lista);
void manipulaElementos(Jogador *lista);
void decodifica(Jogador *lista,char *entrada);
void inserirInicio(Jogador *lista, char* entrada);
void inserir(Jogador *lista, int posicao, char* entrada);
void inserirFim(Jogador *lista, char *entrada);
void removerInicio(Jogador *lista);
void remover(Jogador *lista, int posicao);
void removerFim(Jogador *lista);




void inserirElementos(Jogador *lista){
    char input[100];
    scanf("%s", input);

    while(strcmp(input, "FIM")!= 0){
        int id = atoi(input);
        setaId(&lista[tamanho++],id);
        scanf("%s", input);
    }
}

void setaId(Jogador *jogador, int id) {
    FILE *file;
    file = fopen("/tmp/players.csv", "r");

    char linha[100];
    char resp[100];
    bool achou = false;
    char *token;

    fgets(linha, sizeof(linha), file);

    while (fgets(linha, sizeof(linha), file) != NULL && !achou){
        strcpy(resp,linha);
        token = strtok(linha,",");
        int num = atoi(token);

        if(id==num){
            parteDados(jogador,resp);
            achou=true;
        }
        token=strtok(NULL,",");

    }
    fclose(file);
}

void parteDados(Jogador *jogador, const char *x) {

    int len = strlen(x);
    char linha[len];
    strcpy(linha,x);

    linha[len-1]='\0';
    int j = 0;
    char *nvlinha = malloc(len + 4);


    for (int i = 0; i < len; i++) {
        if (linha[i]==',' && linha[i+1]==',') {
            nvlinha[j]=linha[i];
            nvlinha[j+1]=' ';
            nvlinha[j+2]=',';
            j+=3;
            i++;
        } else {
            nvlinha[j] = linha[i];
            j++;
        }
    }

    if(nvlinha[j-2]==','){
        nvlinha[j-1]=' ';
    }
    nvlinha[j]='\0';

    setaDados(jogador,nvlinha);
    free(nvlinha);
}

void setaDados(Jogador *jogador,char *linha){

    int t=0;
    char *parte[8];
    char *token = strtok(linha, ",");

    while (token != NULL && t < 8) {
        parte[t] = token;
        token = strtok(NULL, ",");          //MANTEM UM ESTADO INTERNO PARA RASTREAR A POSICAO ATUAL NA STRING
        t++;
    }

    jogador->id = atoi(parte[0]);
    if(strcmp(parte[1], " ")== 0)strcpy(jogador->nome, "nao informado");else strcpy(jogador->nome, parte[1]);
    jogador->altura = atoi(parte[2]);
    jogador->peso = atoi(parte[3]);
    if(strcmp(parte[4], " ")== 0)strcpy(jogador->universidade, "nao informado");else strcpy(jogador->universidade, parte[4]);
    jogador->anoNascimento = atoi(parte[5]);
    if(strcmp(parte[6], " ")== 0)strcpy(jogador->cidadeNascimento, "nao informado");else strcpy(jogador->cidadeNascimento, parte[6]);
    if(strcmp(parte[7], " ")== 0)strcpy(jogador->estadoNascimento, "nao informado");else strcpy(jogador->estadoNascimento, parte[7]);

    //mostraDados(jogador);

}

void mostraDados(Jogador *jogador){
    printf("[%i] ## %s ## %d ## %d ## %d ## %s ## %s ## %s ##\n",
        mostrar,
        jogador->nome,
        jogador->altura,
        jogador->peso,
        jogador->anoNascimento,
        jogador->universidade,
        jogador->cidadeNascimento,
        jogador->estadoNascimento
    );
    mostrar++;
}

void mostrarTodos(Jogador *lista){
    for(int i=0;i<tamanho;i++){
        mostraDados(&lista[i]);
    }
}

void manipulaElementos(Jogador *lista){
    int qtd; scanf("%i",&qtd);
    char entrada[20];

    while (getchar() != '\n');

    for (int i=0;i<qtd;i++){
        fgets(entrada, sizeof(entrada), stdin);                     // O FGETS COLOCA UM "\N" NO FINAL DA STRING;
        decodifica(lista,entrada);
    }
}

void decodifica(Jogador *lista,char *entrada){
    int len = strlen(entrada);
    if (len>0 && entrada[len-1]=='\n'){entrada[len-1]='\0';}
    char *token = strtok(entrada," ");

    if(token!=NULL){
        if (strcmp(token,"II")==0){
            token = strtok(NULL, " ");
            inserirInicio(lista,token);
        }
        else if(strcmp(token, "IF")==0){
            token = strtok(NULL, " ");
            inserirFim(lista,token);
        }
        else if(strcmp(token,"I*")==0){
            token = strtok(NULL, " ");
            int posicao = atoi(token);
            token = strtok(NULL, " ");
            inserir(lista,posicao,token);
        }
        else if(strcmp(token,"R*")==0){
            token = strtok(NULL, " ");
            int posicao = atoi(token);
            remover(lista,posicao);
        }
        else{
            if(strcmp(entrada, "RI")==0){
                removerInicio(lista);
            }
            else if(strcmp(entrada,"RF")==0){
                removerFim(lista);
            }
        }
    }
}

void inserirInicio(Jogador *lista, char* entrada){
    int idJogador = atoi(entrada);                          /*ATOI ESTA COLOCANDO UM \N NO ID JOGADOR*/
    for (int i=tamanho;i>0;i--){
       lista[i]=lista[i-1];
    }
    setaId(&lista[0],idJogador);
    tamanho++;
}

void inserir(Jogador *lista, int posicao, char* entrada){
    if(posicao==0) inserirInicio(lista,entrada);
    else if(posicao==tamanho) inserirFim(lista,entrada);
    else{
        int idJogador = atoi(entrada);
        for (int i=tamanho;i>posicao;i--){
            lista[i]=lista[i-1];
        }
        setaId(&lista[posicao],idJogador);
        tamanho++;
    }
}

void inserirFim(Jogador *lista, char *entrada){
    int idJogador = atoi(entrada);
    setaId(&lista[tamanho],idJogador);
    tamanho++;
}

void removerInicio(Jogador *lista){
    mostraRemovido(&lista[0]);

    for (int i=0;i<tamanho;i++){
        lista[i]=lista[i+1];
    }
    tamanho--;
}

void remover(Jogador *lista, int posicao){
    mostraRemovido(&lista[posicao]);

    if(posicao==0) removerInicio(lista);
    else if(posicao==tamanho-1) removerFim(lista);
    else{
        for (int i=posicao;i<tamanho-1;i++){
            lista[i]=lista[i+1];
        }
    }
    tamanho--;
}

void removerFim(Jogador *lista){
    mostraRemovido(&lista[tamanho-1]);
    tamanho--;
}

void mostraRemovido(Jogador *jogador){
    printf("(R) %s\n", jogador->nome);
}

int main(){
    Jogador lista[500];
    inserirElementos(lista);
    manipulaElementos(lista);
    mostrarTodos(lista);
    return 0;
}












